main, test: integrate perf tools into scylla

tchaikov opened the pull request:
following tests are integrated into scylla executable

- perf_fast_forward
- perf_row_cache_update
- perf_simple_query
- perf_row_cache_update
- perf_sstable

before this change
```console
$ size build/release/scylla
   text    data     bss     dec     hex filename
82284664         288960  335897 82909521        4f11951 build/release/scylla
$ ls -l build/release/scylla
-rwxrwxr-x 1 kefu kefu 1719672112 Jan 19 17:51 build/release/scylla
```
after this change
```console
$ size build/release/scylla
   text    data     bss     dec     hex filename
84349449         289424  345257 84984130        510c142 build/release/scylla
$ ls -l build/release/scylla
-rwxrwxr-x 1 kefu kefu 1774204800 Jan 19 17:52 build/release/scylla
```

Fixes #12484

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3894/

denesb commented on main.cc:
```diff
@@ -1740,30 +1740,42 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+        {"types", tools::scylla_types_main},
+        {"sstable", tools::scylla_sstable_main},
+    };
+    auto found = std::ranges::find_if(funcs, [name](auto& name_and_func) {
```
@avikivity can we now use ranges?

denesb replied:
Please add space between `]` and `(`.

tchaikov replied:
@denesb hi Botond, we are already using them =)
```console
$ git grep std::ranges | wc -l
45
```
BTW, i recall that Avi asked the same question when i was using ranges in Seastar!

denesb replied:
I once started sending patches converting us from `boost::ranges` to `std::ranges`, then we had to revert these due to some bugs in the standard library. Maybe these bugs were fixed, but every time I see `std::range` I think of this.

tchaikov replied:
i see. that's sad. probably the new gcc in the toolchain brought a more C++20 compliant libstdc++ to us.

denesb replied:
The bug might have been fixed since.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3895/

denesb commented on configure.py:
```diff
@@ -1151,6 +1151,7 @@ def find_headers(repodir, excluded_dirs):
 
 scylla_tests_generic_dependencies = [
     'test/lib/cql_test_env.cc',
+    'test/lib/debug.cc'
```
This should be just `debug.cc` (we already have `debug.hh`).

tchaikov commented:
@avikivity thanks for your review. shall i add all `test/perf/*` tests to scylla?

```
ninja: error: 'test/lib/debug.cctest/lib/test_services.cc', needed by 'build/debug/test/lib/debug.otest/lib/test_services.o', missing and no known rule to make it
```

it builds and runs fine locally with ` ./configure.py --with scylla --mode debug` though. will take a closer look at the FTBFS tomorrow.

denesb commented on configure.py:
```diff
@@ -1899,7 +1905,9 @@ def query_seastar_flags(pc_file, link_static_cxx=False):
             else:
-                f.write('build $builddir/{}/{}: {}.{} {} | {}\n'.format(mode, binary, regular_link_rule, mode, str.join(' ', objs), seastar_dep))
+                if binary == 'scylla':
+                    local_libs += ' ' + "$seastar_testing_libs_{}".format(mode)
```
Not thrilled about adding all this test stuff to scylla.
Maybe we should look into loosing the test dependency first.

denesb replied:
Might not be worth it. If it is too complicated, we can defer it to a follow-up.

tchaikov replied:
we are using `src/testing/random.cc` which is compiled in `seastar_testing`.

tchaikov replied:
> Not thrilled about adding all this test stuff to scylla.
> Maybe we should look into loosing the test dependency first.

i did look into this. no luck.

tchaikov replied:
i added more details in the commit message. HTH.

avikivity commented:
> @avikivity thanks for your review. shall i add all `test/perf/*` tests to scylla?
> 


I think so, maybe not all are important, but many are.


/cc @tgrabiec

denesb reviewed (changes requested):
Please check binary size before/after. I don't expect big change but lets check.

tchaikov commented:
changelog

- add "/debug.cc` instead.
- add space in between `]` and `(`.

note to myself, i will 

- [ ] find out the diff in size
- [ ] fix FTBFS
- [ ] explain the new dependency of `seastar_testing` in commit message.

avikivity commented:
> > Please check binary size before/after. I don't expect big change but lets check.
> 
> @denesb Hi Botond, thank you for your review. after a second thought, do we have a dedicated build just for testing? so we don't need to worry about the bloating size? i don't really like to have so many "features" only used for testing delivered to user in production.

We don't, and as soon as we integrate profile-guided optimizations, then we have to use a single build for both testing and production, so we test the same binary.

tchaikov commented:
changelog

- fixed the FTBFS (added a missing comma in `configure.py`)
- add "/debug.cc` instead.
- add space in between `]` and `(`.
- added two more perf tests to scylla
  -  `perf_row_cache_update`
  - `perf_sstable`
 - note down the size of `scylla` before and after the change in commit mesage

denesb commented on configure.py:
```diff
@@ -1172,9 +1171,17 @@ def find_headers(repodir, excluded_dirs):
+                'seastar/tests/perf/linux_perf_event.cc']
 
 deps = {
-    'scylla': idls + ['main.cc', 'debug.cc'] + scylla_core + api + alternator + redis + scylla_tools,
```
and neither this

denesb reviewed (approved):
LGTM, one minor comment only.

denesb commented:
> * note down the size of `scylla` before and after the change in commit mesage

No need to add this to each commit message, it is enough to add to the cover-letter only and compare scylla before vs. scylla after the series.

tchaikov commented on configure.py:
```diff
@@ -1173,7 +1174,7 @@ def find_headers(repodir, excluded_dirs):
 
 deps = {
-    'scylla': idls + ['main.cc'] + scylla_core + api + alternator + redis + scylla_tools,
+    'scylla': idls + ['main.cc', 'debug.cc'] + scylla_core + api + alternator + redis + scylla_tools,
```
yes, that's even better!

avikivity commented:
> > * note down the size of `scylla` before and after the change in commit mesage
> 
> No need to add this to each commit message, it is enough to add to the cover-letter only and compare scylla before vs. scylla after the series.

Better: the output of `size scylla` before and after, so we can see how it looks after debug information is stripped.

Though it's not so important, the tests are tiny compared the scylla core.

tchaikov commented:
changelog

- note down the size before and after the changeset in the cover letter
- move `debug.cc` to `scylla_core`

tchaikov commented:
> > > * note down the size of `scylla` before and after the change in commit mesage
> > 
> > 
> > No need to add this to each commit message, it is enough to add to the cover-letter only and compare scylla before vs. scylla after the series.
> 
> Better: the output of `size scylla` before and after, so we can see how it looks after debug information is stripped.

added.

avikivity commented:
It added 2MB, more than I expected. Still, it's reasonable, and worthwhile as we can now easily run the performance tests on any machine using packages.

michoecho commented:
> We don't, and as soon as we integrate profile-guided optimizations, then we have to use a single build for both testing and production, so we test the same binary.

You mean link-time optimizations (and/or BOLT), not profile-guided optimizations. With LTO disabled, all executables are linked from the same object files. (Not that it matters, because we aren't planning to use PGO without LTO).

avikivity commented:
> > We don't, and as soon as we integrate profile-guided optimizations, then we have to use a single build for both testing and production, so we test the same binary.
> 
> You mean link-time optimizations (and/or BOLT), not profile-guided optimizations. With LTO disabled, all executables are linked from the same object files. (Not that it matters, because we aren't planning to use PGO without LTO).

Yes

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3901/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3903/
